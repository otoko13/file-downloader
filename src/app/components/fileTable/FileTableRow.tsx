import { FunctionComponent, useMemo } from "react";
import Checkbox from "../checkbox/Checkbox";
import { DownloadableFile } from "@/types";
import AvailableMarker from "../availableMarker/AvailableMarker";
import styles from "./fileTableRow.module.scss";
import classNames from "classnames";

interface FileTableRowProps {
  file: DownloadableFile;
  selected: boolean;
  onClick: () => void;
}

const FileTableRow: FunctionComponent<FileTableRowProps> = ({
  file,
  selected,
  onClick,
}) => {
  const statusText = useMemo(() => {
    switch (file.status) {
      case "available":
        return "Available";
      case "scheduled":
        return "Scheduled";
      default:
        return "";
    }
  }, [file.status]);

  return (
    <tr
      className={classNames(styles["file-table-row"], {
        [styles.selected]: selected,
      })}
      onClick={onClick}
    >
      <td className="selected">
        <Checkbox checkedState={selected ? "checked" : "unchecked"} />
      </td>
      <td>{file.name}</td>
      <td>{file.device}</td>
      <td>{file.path}</td>
      <td>
        <div className={styles["status-cell"]}>
          <AvailableMarker available={file.status === "available"} />
          {statusText}
        </div>
      </td>
    </tr>
  );
};

export default FileTableRow;
