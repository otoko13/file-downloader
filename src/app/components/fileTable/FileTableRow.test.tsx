import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import FileTableRow, { FileTableRowProps } from "./FileTableRow";

let defaultProps: FileTableRowProps;
let clickHandler: jest.Mock;

const renderTableWithRow = (props: FileTableRowProps) =>
  render(
    <table>
      <tbody>
        <FileTableRow {...props} />
      </tbody>
    </table>
  );

beforeEach(() => {
  defaultProps = {
    file: {
      name: "smss.exe",
      device: "Mario",
      path: "\\Device\\HarddiskVolume2\\Windows\\System32\\smss.exe",
      status: "scheduled",
    },
    selected: false,
    onClick: clickHandler,
  };
});

describe("appearance", () => {
  it("should match the snapshot", () => {
    const { container } = renderTableWithRow(defaultProps);
    expect(container).toMatchSnapshot();
  });
});

describe("checkbox state", () => {
  it("should not be checked if the row is not selected", () => {
    renderTableWithRow(defaultProps);
    const checkbox = screen.getByRole("checkbox");
    expect(checkbox.getAttribute("aria-checked")).toBe("false");
  });

  it("should be checked if the row is selected", () => {
    renderTableWithRow({ ...defaultProps, selected: true });
    const checkbox = screen.getByRole("checkbox");
    expect(checkbox.getAttribute("aria-checked")).toBe("true");
  });
});

describe("available marker", () => {
  it("should not be shown if the file is not available", () => {
    renderTableWithRow(defaultProps);
    expect(screen.getByLabelText("available")).not.toHaveClass("visible");
  });

  it("should be shown if the file is available", () => {
    renderTableWithRow({
      ...defaultProps,
      file: { ...defaultProps.file, status: "available" },
    });
    expect(screen.getByLabelText("available")).toHaveClass("visible");
  });

  it("should show Scheduled if scheduled", () => {
    renderTableWithRow(defaultProps);
    expect(screen.queryByText("Scheduled")).toBeInTheDocument();
  });

  it("should show Available if available", () => {
    renderTableWithRow({
      ...defaultProps,
      file: { ...defaultProps.file, status: "available" },
    });
    expect(screen.queryByText("Available")).toBeInTheDocument();
  });

  it("should show nothing if neither scheduled nor available", () => {
    renderTableWithRow({
      ...defaultProps,
      file: { ...defaultProps.file, status: "" as "available" },
      // need to do the cast to allow the bad value
    });
    expect(screen.queryByText("Available")).not.toBeInTheDocument();
    expect(screen.queryByText("Scheduled")).not.toBeInTheDocument();
  });
});
