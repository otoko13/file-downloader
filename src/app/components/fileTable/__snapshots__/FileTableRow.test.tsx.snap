// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appearance should match the snapshot 1`] = `
<div>
  <table>
    <tbody>
      <tr
        class="file-table-row"
      >
        <td
          class="selected"
        >
          <button
            aria-checked="false"
            aria-label="Select"
            class="checkbox"
            role="checkbox"
          />
        </td>
        <td>
          smss.exe
        </td>
        <td>
          Mario
        </td>
        <td>
          \\Device\\HarddiskVolume2\\Windows\\System32\\smss.exe
        </td>
        <td>
          <div
            class="status-cell"
          >
            <div
              aria-hidden="true"
              aria-label="available"
              class="available-marker"
              role="img"
            />
            Scheduled
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
